README.rst
============================
Django Agnostic Autocomplete
============================


.. image:: https://travis-ci.org/peopledoc/django-agnocomplete.svg?branch=master
    :target: https://travis-ci.org/peopledoc/django-agnocomplete


Heavily based on `django-autocomplete-light v2 <https://github.com/yourlabs/django-autocomplete-light/>`_ workflow and concepts, this toolkit offers a front-end agnostic way to get fields for autocompletion.

It will provide:

* a simple and configurable entry-point management,
* a REST-like HTTP API to search for results,
* Fields and widgets that will make the interface between our Django code and *your* Javascript.

Status
======

Stable, used in production.

Install
=======

.. code:: sh

    $ pip install django-agnocomplete

Or add ``django-agnocomplete`` to your project requirements.

Documentation
=============

`The full documentation is browsable on Read the Docs <http://django-agnocomplete.readthedocs.org/en/latest/>`_


Tests
=====

Install ``tox`` in your environment (it could be a virtualenv) and run:

.. code:: sh

    $ tox

It'll run the tests for all the combinations of the following:

* Python 3.6, 3.7, 3.8, 3.9.
* Django 2.2.

and a ``flake8`` check.

Are you a developer?
--------------------

To target a specific test case, use the following:

.. code:: sh

    $ tox -e py37-django22 --  demo.tests.test_core.AutocompleteChoicesPagesOverrideTest

Everything after the double-dash will be passed to the django-admin.py test command.

If you need to install a debugger (let's say `ipdb`), you can use the ``TOX_EXTRA`` environment variable like this:

.. code:: sh

    $ TOX_EXTRA=ipdb tox -e py36-django22

.. note::

    We've got a self documented Makefile for common tasks, such as running the tests, building the docs, etc.

Run the demo
============

The (draft) demo site can be browsed using the Django devserver. Run:

.. code:: sh

    $ make serve

It will run a syncdb (it may ask you questions) and then a runserver with your current ``demo.settings``. You can browse the (very rough) website at http://127.0.0.1:8000/. You can add
any runserver options you want using the `tox` positional parameters, like this:

.. code:: sh

    $ tox -e serve -- 9090  # to change the listening port


Here you'll be able to see that ``django-agnocomplete`` has been easily and rapidly integrated with ``selectize.js``, ``select2``, ``jquery-autocomplete`` and ``typeahead``. With the same backend, you can plug the JS front-end you want.

Troubles running the demo?
--------------------------

This demo project is not build as a production-ready application, models can change, but there's no migration in it. If you have database errors, you can try to remove it using:

.. code:: sh

    $ make clean-db

Or, for more radical cleanup:

.. code:: sh

    $ make clean-all


License
=======

This piece of software is being published under the terms of the MIT License. Please read the `LICENSE` file for more details.


CHANGELOG.rst
=================================
Changelog for django-agnocomplete
=================================

master (unreleased)
===================

Nothing here yet.

2.2.0 (2022-04-21)
==================

* Fix path warnings (#132) @alyohea
* Publish workflow should just run on tag push (#131) @ewjoachim
* Create CODEOWNERS (#130) @ewjoachim
* Drop six (#129) @alyohea

2.1.0 (2022-01-14)
==================

* Adding missing pyproject.toml, needed to automatically find the version (#128) @mike-perdide
* Remove version from setup.cfg to let the GH publish action do its magic (#127) @mike-perdide
* Update tox.ini and enable Github Actions, focusing on compatibilty with Django 3.x to 4.0 (#125) @mike-perdide
* Attempt to make it installable with Django 3.X serie (#124) @yohanboniface

2.0.0 (2021-03-26)
==================

* Confirm compatibility with Python 3.9.
* Drop support for Django 2.0, 2.1 (#116).
* Drop support for Python 3.5 (#114).
* Drop support for Django 1.11 & Python 2.7 (#115).

1.0.0 (2020-01-29)
==================

* Cleanup code related to Django 1.8, 1.9 & 1.10
* Remove tox targets for Django 1.8, 1.9 & 1.10
* Add support for Django 2.0, 2.1 & 2.2
* Fixed Demo site.
* Cleaned up README and other docs reflecting the newest compatibilities.

0.14.0 (2018-10-16)
===================

* Add a way in UrlProxy widget to filter value with python (#104)
* Provide a FieldMixin in order to use with UrlProxy Autocomplete for efficient value validation (#107)

0.13.0 (2018-10-02)
===================

* Added ``doclint`` & ``flake8`` to travis builds (#99).
* Fixed numerous flake8 errors (#99).
* Dropped support for Django 1.8, 1.9, 1.10 (#98).
* Added support for python 3.6
* Fixed User.is_authenticated() deprecation warning (Django>=2.0)
* Fixed deprecated MIDDLEWARE_CLASSES settings
* Fixed deprecated urls declarations
* Fixed ForeignKey on_delete values

0.12.2 (2018-03-09)
===================

* Fix AgnocompleteWidgetMixin running unfiltered queries on django>=1.11

0.12.1 (2018-03-08)
===================

**THIS VERSION IS CANCELLED**. We messed up a bit with branches and releases, and it resulted in a bad release.
Please consider that this release has never existed.

We're deeply sorry about the inconvenience.

0.12.0 (2018-01-26)
===================

* Fix bug in AgnocompleteWidgetMixin when template-based widgets are used (Django>=1.11).

0.11.0 (2018-01-25)
===================

* Various Flake8 fixes (#89).
* Fix the multiple Model Field search with authentication behaviour. Please note that this bug would only occur if ``Django<1.11`` (#91).

0.10.0 (2018-01-10)
==================

- Add Django 1.11 support (#85).
- Moving Github organization from ``novafloss`` to ``peopledoc`` (#86).

0.9.0 (2017-07-11)
==================

- Added a more flexible way to create new values in AgnocompleteModelMultipleField (#82).

0.8.0 (2017-03-27)
==================

- Drop Python 3.3 support (#81).
- Add Django 1.10 support (#80).


0.7.0 (2016-12-13)
==================

Enhancements & New Features
---------------------------

- Return the eventual HTTP error message to the front-end in the context of a HTTP error in a ``AgnocompleteURLProxy`` field (#71).
- Handle the ``to_field_name`` parameter with ``AgnocompleteModel`` and allow customization of the label alone by overriding ``AgnocompleteModel.label()`` (#77).
- In the "error" demo, display the error message returned by the Agnocomplete call (#65).

Docs & Help
-----------

- Added a ``make clean`` command to remove junk assets (#64).
- Added a ``doclint`` job to check documentation build (#69).
- Link the Github project on the documentation homepage (#73).
- Mention the version of `django-autocomplete-light` it reuses concepts from (#74).
- Update README (typos, syntax HL on commands) (#75).

0.6.0 (2016-10-10)
==================

- Dropped support for Django 1.6 / 1.7 (#54),
- Added support for Django 1.9. Please note that the combination Python 3.3 and Django 1.9 is incompatible - `see Django 1.9 release notes <https://docs.djangoproject.com/en/1.10/releases/1.9/>`_ (#56).
- Added support for extra arguments passed to the search URL, passed on the Agnocomplete class (#52).
- Added the ``AgnocompleteUrlProxy`` class, handling autocomplete using a third-party HTTP API (#55, #62, #63, #67).
- Removed Django 1.10 deprecation warnings (#59).
- Global Error Handling (#60).
- Allowing Autocomplete class argument in AgnocompleteField to be either string (``str``) or unicode variables (#66).

0.5.0 (2016-07-01)
==================

- Removed Django deprecation (#49)
- Now ready for Python 3.5. (#19) - Note: Only available for Django 1.8 and above.

0.4.0 (2016-02-04)
==================

- Added the multiple selection feature (#33).


0.3.2 (2016-01-27)
==================

- added a new method in ``AgnocompleteModel``, named ``build_filtered_queryset``, to allow overriding (#47).


0.3.1 (2015-12-04)
==================

- Fix IE8/9 bug for AJAX response headers (#45)


0.3.0 (2015-11-06)
==================

- Stronger validation of context-based agnocomplete fields (#39).
- Expose a ``final_queryset`` (aliasing the ``_final_queryset`` property) *and* a ``final_raw_queryset`` property that recieves the actual *unpaginated* queryset on which the search is based (#40).


0.2.3 (2015-11-05)
==================

- Expose a ``_final_queryset`` property that receives the actual queryset executed right before serialization (#40).


0.2.2 (2015-10-12)
==================

- Improve performances by slicing the resultset before rendering (#36).
- Added an `item(current_item)` method to override display label on choices (#37).


0.2.1 (2015-09-30)
==================

- Add a new method to have the possibilty to override easily the display label (#34).
- ``make docs`` is a PHONY makefile target.

0.2.0 (2015-09-17)
==================

Feature(s)
----------

- a more pertinent data attribute to target agnocomplete-ready fields (#22).

demo/settings.py
"""
Django settings for demo project.

Generated by 'django-admin startproject' using Django 1.8.4.

For more information on this file, see
https://docs.djangoproject.com/en/1.8/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.8/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.8/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '9%pjcpe2x*39$0+ci*2y@x9du@^&yl%woc+f#+%!yen4@n_nx1'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    # agnostic autocomplete
    'agnocomplete',

    # Django contribs
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    # demo app
    'demo',
)

ROOT_URLCONF = 'demo.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]


# Database
# https://docs.djangoproject.com/en/1.8/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

MIDDLEWARE = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

# Internationalization
# https://docs.djangoproject.com/en/1.8/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.8/howto/static-files/

STATIC_URL = '/static/'

TEST_RUNNER = 'django.test.runner.DiscoverRunner'

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {
        'require_debug_false': {
            '()': 'django.utils.log.RequireDebugFalse'
        }
    },
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'filters': ['require_debug_false'],
            'class': 'django.utils.log.AdminEmailHandler'
        },
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
        },
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
        'django.db.backends': {
            'level': 'INFO',
            # 'level': 'DEBUG',  # for more output
            'handlers': ['console'],
        },
        'agnocomplete': {
            'level': 'DEBUG',
            'handlers': ['console'],
        },
        'demo': {
            'level': 'INFO',
            'handlers': ['console'],
        }
    }
}

AUTHENTICATION_BACKENDS = (
    'demo.authentication.PersonBackend',
    'django.contrib.auth.backends.ModelBackend',
)

# AGNOCOMPLETE specifics
# 1. Page size
AGNOCOMPLETE_DEFAULT_PAGESIZE = 15
AGNOCOMPLETE_MAX_PAGESIZE = 120
AGNOCOMPLETE_MIN_PAGESIZE = 2
# 2. Query size
AGNOCOMPLETE_DEFAULT_QUERYSIZE = 5
AGNOCOMPLETE_MIN_QUERYSIZE = 4
# 3. agnocomplete prefix
AGNOCOMPLETE_NAMESPACE = 'autocomplete'

# Demo specifics
HTTP_HOST = 'http://127.0.0.1:8000'
